velox/exec/MergeJoin.cpp

```diff
@@ -261,13 +270,34 @@ void MergeJoin::addOutputRowForLeftJoin(
   ++outputSize_;
 }
 
+void MergeJoin::flattenRightProjections() {
+  auto& children = output_->children();
+
+  for (const auto& projection : rightProjections_) {
+    auto& currentVector = children[projection.outputChannel];
+    auto newFlat = BaseVector::create(
+        currentVector->type(), outputBatchSize_, operatorCtx_->pool());
+    newFlat->copy(currentVector.get(), 0, 0, outputSize_);
+    children[projection.outputChannel] = newFlat;
+  }
+  isRightFlattened_ = true;
+}
+
 void MergeJoin::addOutputRow(
     const RowVectorPtr& left,
     vector_size_t leftIndex,
     const RowVectorPtr& right,
     vector_size_t rightIndex) {
-  copyRow(left, leftIndex, output_, outputSize_, leftProjections_);
-  copyRow(right, rightIndex, output_, outputSize_, rightProjections_);
+  // All left side projections (and the right side) share the same dictionary
+  // indices.
+  rawLeftIndices_[outputSize_] = leftIndex;
+
+  // If right side projections are not a dictionary, do a deep copy.
```

Comment:
Is this comment accurate? I'm not sure I understand it.

Reply:
I can improved the comments, but the point is that adding a row is just adding the dictionary index, unless the right projections have been flattened. In that case, you actually need to copy the value out.  The idea being that most of the time you would not hit the copy.